### PR TITLE
Add Undo and Redo commands

### LINKED ISSUE #85: As a user, I want to be able to undo and redo so that I can easily fix mistakes when using the application

### PR DESCRIPTION
**Pull request by bchenghi**
- Add Undo and Redo commands
- Add VersionedFoodInventory that stores food inventory states
- Add InvalidationListenerManager for event handling (currently between VersionedFoodInventory and LogicManager)
- Add Undo, Redo, VersionedFoodInventory and InvalidationListenerManager tests
- Update DG and UG with undo and redo
- Add assertions to VersionedFoodInventory, UndoCommand and RedoCommand
- Add credits to VersionedFoodInventory, UndoCommand, RedoCommand, InvalidationListenerManager and their tests
- Add logging to VersionedFoodInventory

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -180,17 +177,17 @@ The following sequence diagram shows how the undo operation works:
...
+<div markdown="span" class="alert alert-info">:information_source: **Note:** If the `currentStatePointer` is at index `foodInventoryStateList.size() - 1`, pointing to the latest food inventory state, then there are no undone food inventory states to restore. The `redo` command uses `Model#canRedoFoodInventory()` to check if this is the case. If so, it will return an error to the user rather than attempting to perform the redo.
 
 </div>
 
-Step 5. The user then decides to execute the command `list`. Commands that do not modify the address book, such as `list`, will usually not call `Model#commitAddressBook()`, `Model#undoAddressBook()` or `Model#redoAddressBook()`. Thus, the `addressBookStateList` remains unchanged.
+Step 5. The user then decides to execute the command `list`. Commands that do not modify the food inventory, such as `list`, will usually not call `Model#commitFoodInventory()`, `Model#undoFoodInventory()` or `Model#redoAddressBook()`. Thus, the `foodInventoryStateList` remains unchanged.
```
Comment (jonahtanjz) [inline docs/DeveloperGuide.md]: `Model#redoAddressBook()` should change to FoodInventory

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -133,41 +133,38 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
-The proposed undo/redo mechanism is facilitated by `VersionedAddressBook`. It extends `AddressBook` with an undo/redo history, stored internally as an `addressBookStateList` and `currentStatePointer`. Additionally, it implements the following operations:
-
-* `VersionedAddressBook#commit()` — Saves the current address book state in its history.
-* `VersionedAddressBook#undo()` — Restores the previous address book state from its history.
-* `VersionedAddressBook#redo()` — Restores a previously undone address book state from its history.
+* `VersionedFoodInventory#commit()` — Saves the current food inventory state in its states list.
+* `VersionedFoodInventory#undo()` — Restores the previous food inventory state from its states list.
+* `VersionedFoodInventory#redo()` — Restores a previously undone food inventory state from its states list.
 
 These operations are exposed in the `Model` interface as `Model#commitAddressBook()`, `Model#undoAddressBook()` and `Model#redoAddressBook()` respectively.
```
Comment (jonahtanjz) [inline docs/DeveloperGuide.md]: Perhaps can change these to FoodInventory?

File: src/main/java/seedu/simplykitchen/model/FoodInventory.java [hunk 3]
```diff
@@ -94,6 +97,23 @@ public void removeFood(Food key) {
...
+        invalidationListenerManager.addListener(listener);
+    }
+
+    @Override
+    public void removeListener(InvalidationListener listener) {
+        invalidationListenerManager.removeListener(listener);
+    }
+
+    /**
+     * Notifies listeners that the address book has been modified.
```
Comment (jonahtanjz) [inline src/main/java/seedu/simplykitchen/model/FoodInventory.java]: JavaDoc contains address book

**Review by jonahtanjz — COMMENTED** [review]
Some traces of address book left otherwise it looks good :)

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (bchenghi) [reply docs/DeveloperGuide.md]: Changed. Thanks!

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (bchenghi) [reply docs/DeveloperGuide.md]: Okay changed. Thanks!

File: src/main/java/seedu/simplykitchen/model/FoodInventory.java (hunk 3 above)
Comment (bchenghi) [reply src/main/java/seedu/simplykitchen/model/FoodInventory.java]: Okay. Thanks!

**Review by jonahtanjz — APPROVED** [review]
LGTM 👍

**Pull request merged**